src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -336,11 +359,75 @@ public void CopyChangedValues(TStructuralType original)
                     bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
                     Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
 
+                    Type nestedType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && !nestedType.IsAssignableFrom(originalType))
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, nestedType, originalType);
+                    }
+
                     deltaNestedResource.CopyChangedValues(originalNestedResource);
                 }
             }
         }
 
+        private dynamic ReAssignComplexDerivedType(TStructuralType original, string nestedResourceName, dynamic originalNestedResource, Type nestedType, Type originalType)
+        {
```

Comment:
This method (and its members) should be named according to what it does, not how how it is used.  i.e., calling parameters "nestedType" and "original" bring in context from the caller that make it difficult to understand locally. Instead, if we just think of this method as reassigning a property, we could have a signature such as:

private dynamic ReAssignComplexProperty(TStructuralType parent, string propertyName, dynamic newValue, Type newType, Type originalType)

This makes it easier to understand this method without understanding how the method is used.